### DISCUSSION TITLE
Have you seen those little dedicated TFT meters?

### DISCUSSION THREAD

**Discussion by TangerineDreamer** [Bodmer/TFT_eSPI#2066] [General]
Look at that:

https://fr.aliexpress.com/item/1005004649714249.html?ug_edm_item_id=1005004649714249&pdp_npi=2%40dis%21EUR%21%E2%82%AC%2018%2C39%21%E2%82%AC%2014%2C71%21%21%21%21%21%40212a793816651791069898906d29b5%2112000029976568340%21edm&edm_click_module=alg_products&tracelog=rowan&rowan_id1=aeug_edm_24687_1_fr_FR_2022-10-07&rowan_msg_id=0144biz_search_product%3A0%3A0_572449438%24fe08540b3708416e81cd3c96e3453abd&ck=in_edm_other&gatewayAdapt=glo2fra

![meters tft](https://user-images.githubusercontent.com/76820158/194697578-693ab213-0b23-4372-8a50-c4d78a034c2d.jpg)

I'm curious what bodmer can explain about that,is this only software or nextion-like tft?
It seems you can do what you want as meters-like,i can imagine what you can show with a 7inch tft :)

**Comment by Bodmer** [comment]
This is quite simple to do. It is similar to the library examples. The meters can be used by users with no programming experience. It appears the signal analogue value has to be digitised elesewhere and then the meter is commanded by USB. I expect the meters could be hacked dependant on the processors etc.

Notice that the needle does not sweep over anything but a background colour, this makes the meter very easy to implement. with a low cost processor.

Incidentally, when I get time I may add more examples to the TFT_eWidget support library.

**Comment by TangerineDreamer** [comment]
You're right,the background color must be used as transparent,but if i take an image from the reality like they do,i have
to color the background of the sketch with the exact color of the image background,but how did they do that?
I've done a meter with an image and erased the needle to let the needle sketch only,but my backgound was not a united color
enough to cover the tickle moving.
Or just an approximation of this color can do the stuff? 
I think many people like me would like to have a nice set of meters,coming from reality :)

**Reply by Bodmer** [reply]
The approach I have used in the past is to use a image editor like Paint, load the image and use the tools to fine the 24 bit colour of a selected pixel. In Paint use the "dropper" to select a pixel, then "Edit colours" to see the RGB value.

**Comment by TangerineDreamer** [comment]
Here are the values found from the backgroud:
RED = 255 , GREEN = 153 , BLUE = 0
how do you declare  TFT_TRANSPARENT with those values?

#define  TFT_TRANSPARENT(255 ,153 ,0) //like that?

**Reply by Bodmer** [reply]
You have to convert to 16 bit colour:

uint16_t background = tft.color565(255, 153, 0);

The simplest approach is to draw the needle in say red,  if the needle moves draw it in red the new position, then draw over the old needle with background, then(since this may erase part of the new needle position) redraw the new needle position in red again. This minimises flicker. This appraoch is used in the analogue meter example.

If you are using a transparent sprite then bear in mind you need to erase the old needle position as well.

**Comment by TangerineDreamer** [comment]
That's finally working,i'm glad.
I have better results with a needle in black or white  ,or red when the background is white.
I've tried that with the clock sketch,but the needles flicker badly  with a image uploaded.
Is it possible,to use images making the needles and move them ,and another image as transparent,i mean using
images over image ,sprite over sprite ,instead of colors over color ?

**Reply by Bodmer** [reply]
You can paste a sprite into another sprite.

Post a complete sketch and I will have a look to see if it can be optimised for rendering.

**Comment by TangerineDreamer** [comment]
For now,i'm trying to make working a sketch that use a PNG file of 32Ko uploaded with LittleFS in a Wemos MiniD1,
at compiling i have an error  " bss' is not within region `dram0_0_seg'"
It's strange because i was able to load a BMP file of 510Ko with the SPIFFS with a MiniD1 !!.
I saw someone on a forum talking about MMU cache : 16Kb + 48Kb IRAM to resolve the problem,but that not work.
I see in the sketch "LittleFS_PNG" to allocate partition with the tools,is it in the flash size menu?

**Reply by Bodmer** [reply]
I suspect ESP8266 does not have enough RAM for PNG decoding, the Pngdec Library uses large RAM buffers.

**Reply by petely** [reply]
Yeah, the pngdec library is simply too big for an ESP8266. I have tried adding that code into one of my projects and I get exactly the same linker error.
However when I comment out all the PNG stuff, it builds and links fine. Here's the output:
```
Executable segment sizes:
ICACHE : 32768           - flash instruction cache 
IROM   : 415352          - code in flash         (default or ICACHE_FLASH_ATTR) 
IRAM   : 27593   / 32768 - code in IRAM          (IRAM_ATTR, ISRs...) 
DATA   : 1512  )         - initialized variables (global, static) in RAM/HEAP 
RODATA : 3660  ) / 81920 - constants             (global, static) in RAM/HEAP 
BSS    : 26328 )         - zeroed variables      (global, static) in RAM/HEAP 
Sketch uses 448117 bytes (42%) of program storage space. Maximum is 1044464 bytes.
Global variables use 31500 bytes (38%) of dynamic memory, leaving 50420 bytes for local variables. Maximum is 81920 bytes.
```
So while the code (without pngdec in it) leaves 50kB free, when it is included the linker fails.
Either the linker cannot handle data of that size (maybe trying to allocate a large block where there isn't enough **contiguous** RAM) or that the library has grown past the advertised 40kB and the documentation hasn't kept up.
I wouldn't like to say, one way or the other. However, the take-away is that I'm not going to spend any more time on this. I'll find another way to load graphics to display on my TFT

**Comment by TangerineDreamer** [comment]
Maybe you're right,but why SPIFFS works with 500ko bitmap in my "vuMeter sketch" ?
If you upload a PNG file with SPIFFS,can you read it with in the sketch "LittleFS_PNG" ?

**Reply by Bodmer** [reply]
It is the PNG decompression process that demands multiple buffers to contain decompression tables and intermediate values. Simply moving a large bitmap image pixel by pixel from FLASH uses a tiny amount RAM  and can even be done with an UNO with 1 Kbyte RAM and a few instructions. As noted in the PNGDEC library ReadMe, "Runs on any MCU with at least 48K of free RAM". In my experience allocating more than 40kbytes to buffers on an ESP8266 can cause problems.

**Comment by TangerineDreamer** [comment]
I believe you ,nodeMCU cannot stand it too,but fortunately i have got some new ESP12F-MOD with 16M flash,and now it's ok.

Now i'm studying the "Rotated_Sprite2" sketch,with a clock face bitmap background in the SPIFFS,
i've pushed only one dial sprite with the needle,my goal is to put 2 other sprites to have hr,Mn,Sec and i wish to incorporate
the clock mechanism to animate them.Because i love those needles look enough real ones.

My problem is to make it clear with the needle because i have difficulties to set the length of the needle,the sketch only propose
to rotate the needle like that:

//---------------------------------------------------------------------------------------
needle.setColorDepth(8);
  needle.createSprite(11, 49); // create the needle Sprite 11 pixels wide by 49 high

  needle.fillSprite(TFT_BLACK);          // Fill with black

  // Define needle pivot point
  uint16_t piv_x = needle.width() / 2;   // x pivot of Sprite (middle)
  uint16_t piv_y = needle.height() - 10; // y pivot of Sprite (10 pixels from bottom)
  needle.setPivot(piv_x, piv_y);         // Set pivot point in this Sprite

  // Draw the red needle with a yellow tip
  // Keep needle tip 1 pixel inside dial circle to avoid leaving stray pixels
  needle.fillRect(piv_x - 1, 2, 3, piv_y + 8, TFT_RED);
  needle.fillRect(piv_x - 1, 2, 3, 5, TFT_YELLOW);
//--------------------------------------------------------------------------------------------------------

i don't want the needle to fill the colored rectangles ,all i want i a needle (color red)  with 
"needle.createSprite(11, 49); "  that i can modify ,but only spin over my clock bitmap,how can i?

**Reply by Bodmer** [reply]
Delete this line:
needle.fillRect(piv_x - 1, 2, 3, 5, TFT_YELLOW);

**Comment by TangerineDreamer** [comment]
yes it's lighter without that line of code,
when will you provide us ,the function-prototypes explanation to build a clock?
